File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        B = nums[::-1]
        for i in range(1, len(nums)):
            nums[i] *= nums[i - 1] or 1
            B[i] *= B[i - 1] or 1
        return max(nums + B)

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.ts
// Time Complexity: O(N)
// Space Complexity: O(1)
function maxProduct(nums: number[]): number {
    let highest = nums[0]
    let lowestSoFar = highest
    let highestSoFar = highest

    for (let i = 1; i < nums.length; i++) {
        const num = nums[i]
        const candidateOne = lowestSoFar * num
        const candidateTwo = highestSoFar * num

        highestSoFar = Math.max(num, candidateOne, candidateTwo)
        lowestSoFar = Math.min(num, candidateOne, candidateTwo)

        highest = Math.max(highestSoFar, highest)
    }

    return highest
};